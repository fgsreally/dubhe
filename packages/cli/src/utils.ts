import { join, resolve } from 'path'
import type { SubConfig } from 'dubhe-lib'
import { getLocalPath, getLocalRecord, getRemoteContent, getTypePathInCache, log } from 'dubhe-lib'
import axios from 'axios'
import { normalizePath } from 'vite'

import fse from 'fs-extra'
import { loadConfig } from 'unconfig'

const root = process.cwd()

export function getPkgName(str: string) {
  if (str.startsWith('@'))
    return str.split('/').splice(0, 2).join('/')

  return str.split('/')[0]
}

export function generateExports(imports: string[]) {
  return imports.reduce((p, c) => {
    const [importVar, importSource] = c.split('--')
    if (importVar === 'default')
      return `${p}import _defaultImport from '${importSource}'\nexport {_defaultImport as default}\n`
    if (importVar === '#css')
      return `${p}import '${importSource}'\n`
    return `${p}export {${importVar}} from '${importSource}'\n`
  }, '')
}
export async function getWorkSpaceConfig() {
  const { config } = await loadConfig({
    sources: [
      // load from `my.config.xx`
      {
        files: 'dubhe.config',
        // default extensions
        extensions: ['ts', 'mts', 'cts', 'js', 'mjs', 'cjs', 'json'],
      },

    ],

    merge: false,
  })
  return config as any
}

export async function getRemoteList() {
  const workspaceConfig = await getWorkSpaceConfig()
  return Object.assign(await getLocalRecord(), workspaceConfig.remote)
}

export async function analyseDep(dubheConfig: SubConfig) {
  const ret = {} as Record<string, Set<string>>
  for (const project in dubheConfig.remote) {
    const remoteConfig = await getRemoteContent(`${dubheConfig.remote[project].url}/core/remoteList.json`)
    if (!remoteConfig)
      continue
    for (const dep in remoteConfig.importsGraph) {
      const pkgName = getPkgName(dep)

      if (!ret[pkgName])
        ret[pkgName] = new Set()
      if (remoteConfig.importsGraph[dep].length === 0)
        ret[pkgName].add(`#css--${dep}`)

      else
        remoteConfig.importsGraph[dep].forEach((item: string) => ret[pkgName].add(`${item}--${dep}`))
    }
  }
  return ret
}

export function getDubheDepJS() {
  return resolve(root, 'dubhe.dep.js')
}

export function isExist(p: string) {
  return fse.existsSync(p)
}

export async function downloadFile(url: string, output: string) {
  try {
    const { data: code } = await axios.get(url)
    await fse.outputFile(
      output,
      typeof code === 'string' ? code : JSON.stringify(code),
    )
  }
  catch (e) {
    log(`get remote content fail--${url}`)
    process.exit(1)
  }
}

export async function installProjectCache(baseUrl: string, files: string[], project: string) {
  for (const file of files) {
    // const cachePath = getCachePath(project, file)
    downloadFile(`${baseUrl}/core/${file}`, getLocalPath(project, file))
  }
}

export async function installProjectTypes(baseUrl: string, project: string) {
  const typesFiles = await getRemoteContent(`${baseUrl}/types/types.json`)
  for (const file of typesFiles) {
    const typesPath = getTypePathInCache(project, file)
    downloadFile(`${baseUrl}/types/${file}`, typesPath)
  }
}

export function updateTsConfig(project: string, fileMap: Record<string, string>) {
  let tsconfig
  const TS_CONFIG_PATH = resolve(
    process.cwd(),
    'tsconfig.dubhe.json',
  )
  try {
    tsconfig = fse.readJSONSync(TS_CONFIG_PATH)
  }
  catch (e) {
    tsconfig = {
      compilerOptions: {
        baseUrl: '.',
        paths: {},
      },
    }
  }

  for (const i in fileMap) {
    const jsPath = normalizePath(`./${join('.dubhe/types', project, fileMap[i])}`).replace(/\.ts$/, '')
    // tsconfig.compilerOptions.paths[`!${project}/${i}.*`] = [jsPath]
    tsconfig.compilerOptions.paths[`dubhe-${project}/${i}`] = [jsPath]
  }
  fse.outputJSON(TS_CONFIG_PATH, tsconfig)
}

