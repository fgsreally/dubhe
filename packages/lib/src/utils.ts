import { dirname, extname, isAbsolute, join, normalize, posix, relative, resolve, sep } from 'path'

import { existsSync, lstatSync, readdirSync, rmdirSync } from 'fs'
import os from 'os'
import type { Color } from 'colors'
import colors from 'colors'

import fse from 'fs-extra'
import cv from 'compare-versions'
import debug from 'debug'
import hash from 'hash-string'
import {
  VIRTUAL_HMR_PREFIX,
} from './common'
// import type {

//   extensionType,
// } from './types'

// export function normalizeFileName(name: string) {
//   return `${name}${extname(name) ? '' : '.js'}`
// }

// export function getModuleName(fileName: string) {
//   return fileName.replace(extname(fileName), '')
// }

// export function getHMRFilePath(i: ModuleNode) {
//   return `/${normalizePath(relative(process.cwd(), i?.file || ''))}`
// }
const windowsSlashRE = /\\/g

export function slash(p: string): string {
  return p.replace(windowsSlashRE, '/')
}
export const isWindows = os.platform() === 'win32'

export function normalizePath(id: string): string {
  return posix.normalize(isWindows ? slash(id) : id)
}

export function resolveURLQuery(url: string) {
  if (!url.startsWith(`/${VIRTUAL_HMR_PREFIX}`))
    return false
  const queryUrl = url.replace(`/${VIRTUAL_HMR_PREFIX}`, '')
  const query = new URLSearchParams(queryUrl)
  return {
    file: query.get('file'),
    dir: query.get('dir') || '',
    types: query.get('types') === 'true',
    project: query.get('project'),
    module: JSON.parse(query.get('module') as string) as string[],
  }
}

const fileSet: string[] = []
let rootDir: string

// work for dts plugin only
export function traverseDic(dirPath: string, cb?: (opt: string[]) => void) {
  if (!rootDir)
    rootDir = dirPath
  fse.readdirSync(dirPath, { withFileTypes: true }).forEach((file) => {
    const filePath = join(dirPath, file.name)
    if (file.isFile())
      fileSet.push(normalizePath(relative(rootDir, filePath)))

    else if (file.isDirectory())
      traverseDic(filePath)
  })
  cb?.(fileSet)
}

export function log(msg: string, color: keyof Color = 'green') {
  // eslint-disable-next-line no-console
  console.log(colors[color](`${colors.cyan('[dubhe]')} ${msg}`))
}

// export function getTypePath(project: string,
//   moduleName: string) {
//   return resolve(CACHE_ROOT, project, moduleName)
// }

// export function resolveExtension(
//   extensions: extensionType[],
//   moduleName: string,
//   basename: string,
// ) {
//   for (const i of extensions) {
//     if (extname(moduleName) === i.key)
//       return i.transform(basename)
//   }
// }

export function getRelatedPath(p: string) {
  return normalizePath(relative(process.cwd(), p || ''))
}

export function addExtension(p: string) {
  return p + (extname(p) === '' ? '.ts' : '')
}

export function getShortHash(str: string) {
  return hash(str).toString(16)
}

export function patchVersion(version1: string, version2: string) {
  return cv.compare(version1, version2, '=')
}

export const debugLog = debug('dubhe')

export function getFormatDate() {
  // const date = new Date()
  // const year = date.getFullYear()
  // const month = date.getMonth()
  // const day = date.getDate()
  // return `${year}-${month < 10 ? `0${month}` : month}-${day < 10 ? `0${day}` : day}`
  return new Date().toLocaleString()
}

export function getPkgName(str: string) {
  if (str.startsWith('@'))
    return str.split('/').splice(0, 2).join('/')

  return str.split('/')[0]
}

export function isNativeObj<T extends Record<string, any> = Record<string, any>>(
  value: T,
): value is T {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export function isRegExp(value: unknown): value is RegExp {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

export function isPromise(value: unknown): value is Promise<any> {
  return (
    !!value
    && typeof (value as any).then === 'function'
    && typeof (value as any).catch === 'function'
  )
}

export function mergeObjects<T extends Record<string, unknown>, U extends Record<string, unknown>>(
  sourceObj: T,
  targetObj: U,
) {
  const loop: Array<{
    source: Record<string, any>
    target: Record<string, any>
    // merged: Record<string, any>
  }> = [
    {
      source: sourceObj,
      target: targetObj,
      // merged: mergedObj
    },
  ]

  while (loop.length) {
    const { source, target } = loop.pop()!

    Object.keys(target).forEach((key) => {
      if (isNativeObj(target[key])) {
        if (!isNativeObj(source[key]))
          source[key] = {}

        loop.push({
          source: source[key],
          target: target[key],
        })
      }
      else if (Array.isArray(target[key])) {
        if (!Array.isArray(source[key]))
          source[key] = []

        loop.push({
          source: source[key],
          target: target[key],
        })
      }
      else {
        source[key] = target[key]
      }
    })
  }

  return sourceObj as T & U
}

export function ensureAbsolute(path: string, root: string) {
  return path ? (isAbsolute(path) ? path : resolve(root, path)) : root
}

export function ensureArray<T>(value: T | T[]) {
  return Array.isArray(value) ? value : value ? [value] : []
}

export async function runParallel<T>(
  maxConcurrency: number,
  source: T[],
  iteratorFn: (item: T, source: T[]) => Promise<any>,
) {
  const ret: Promise<any>[] = []
  const executing: Promise<any>[] = []

  for (const item of source) {
    const p = Promise.resolve().then(() => iteratorFn(item, source))

    ret.push(p)

    if (maxConcurrency <= source.length) {
      const e: Promise<any> = p.then(() => executing.splice(executing.indexOf(e), 1))

      executing.push(e)

      if (executing.length >= maxConcurrency)
        await Promise.race(executing)
    }
  }

  return Promise.all(ret)
}

const speRE = /[\\/]/

export function queryPublicPath(paths: string[]) {
  if (paths.length === 0)
    return ''

  else if (paths.length === 1)
    return dirname(paths[0])

  let publicPath = normalize(dirname(paths[0])) + sep
  let publicUnits = publicPath.split(speRE)
  let index = publicUnits.length - 1

  for (const path of paths.slice(1)) {
    if (!index)
      return publicPath

    const dirPath = normalize(dirname(path)) + sep

    if (dirPath.startsWith(publicPath))
      continue

    const units = dirPath.split(speRE)

    if (units.length < index) {
      publicPath = dirPath
      publicUnits = units
      continue
    }

    for (let i = 0; i <= index; ++i) {
      if (publicUnits[i] !== units[i]) {
        if (!i)
          return ''

        index = i - 1
        publicUnits = publicUnits.slice(0, index + 1)
        publicPath = publicUnits.join(sep) + sep
        break
      }
    }
  }

  return publicPath.slice(0, -1)
}

export function removeDirIfEmpty(dir: string) {
  if (!existsSync(dir))
    return

  let onlyHasDir = true

  for (const file of readdirSync(dir)) {
    const abs = resolve(dir, file)

    if (lstatSync(abs).isDirectory()) {
      if (!removeDirIfEmpty(abs))
        onlyHasDir = false
    }
    else {
      onlyHasDir = false
    }
  }

  if (onlyHasDir)
    rmdirSync(dir)

  return onlyHasDir
}

export function debounce(fn: Function, time = 2000) {
  let timer: NodeJS.Timeout
  return (parms?: any) => {
    if (timer)
      clearTimeout(timer)

    timer = setTimeout(() => {
      fn(parms)
    }, time)
  }
}

