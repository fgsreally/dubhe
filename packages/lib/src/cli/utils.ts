import { resolve } from 'path'

import axios from 'axios'

import fse from 'fs-extra'
import { loadConfig } from 'unconfig'
import { getLocalPath, getLocalRecord, getRemoteContent, getTypePathInCache } from '../cache'
import type { PubListType, SubConfig, SubListType } from '../types'
import { getPkgName, log } from '../utils'
import { removeHash } from '../core'

const root = process.cwd()

export function generateExports(imports: string[]) {
  return imports.reduce((p, c) => {
    const [importVar, importSource] = c.split('--')
    if (importVar === '*')
      return `export * from '${importSource}'\n`
    if (importVar === 'default')
      return `${p}import _defaultImport from '${importSource}'\nexport {_defaultImport as default}\n`
    if (importVar === '#side_effect')
      return `${p}import '${importSource}'\n`
    return `${p}export {${importVar}} from '${importSource}'\n`
  }, '')
}
export async function getWorkSpaceConfig() {
  const { config } = await loadConfig({
    sources: [
      {
        files: 'dubhe.config',
        extensions: ['ts', 'mts', 'cts', 'js', 'mjs', 'cjs', 'json'],

      },

    ],

    merge: false,
  })
  return config as any
}

export async function getDubheList() {
  const workspaceConfig = await getWorkSpaceConfig()
  return Object.assign(await getLocalRecord(), workspaceConfig.remote)
}

export async function analysePubDep(dubheConfig: SubConfig, subList?: Record<string, Set<string>>) {
  const ret: Record<string, Set<string>> = subList || {}
  for (const project in dubheConfig.remote) {
    const { url } = dubheConfig.remote[project]
    const pubList: PubListType = await getRemoteContent(`${url}/core/dubheList.json`)
    const subList: SubListType = await getRemoteContent(`${url}/core/dubheList.sub.json`).catch(() => null)
    if (!pubList)
      continue

    for (const dep in pubList.importsGraph) {
      const pkgName = getPkgName(dep)

      if (!ret[pkgName])
        ret[pkgName] = new Set()
      if (pubList.importsGraph[dep].length === 0)
        ret[pkgName].add(`#side_effect--${dep}`)

      else
        pubList.importsGraph[dep].forEach((item: string) => ret[pkgName].add(`${item}--${dep}`))
    }
    if (subList) {
      for (const chain of subList.chains) {
        for (const dep in chain.importsGraph) {
          const pkgName = getPkgName(dep)

          if (!ret[pkgName])
            ret[pkgName] = new Set()
          if (pubList.importsGraph[dep].length === 0)
            ret[pkgName].add(`#side_effect--${dep}`)

          else
            pubList.importsGraph[dep].forEach((item: string) => ret[pkgName].add(`${item}--${dep}`))
        }
      }
    }
  }
  return ret
}
export async function analyseSubDep(subListPath: string) {
  const ret = {} as Record<string, Set<string>>
  try {
    const subList = await fse.readJSON(subListPath)
    for (const dep in subList.importsGraph) {
      const pkgName = getPkgName(dep)

      if (!ret[pkgName])
        ret[pkgName] = new Set()
      if (subList.importsGraph[dep].length === 0)
        ret[pkgName].add(`#side_effect--${dep}`)

      else
        subList.importsGraph[dep].forEach((item: string) => ret[pkgName].add(`${item}--${dep}`))
    }
    return ret
  }
  catch (e) {
    console.log(e)
    log('can\'t find dubheList.json', 'red')
  }
}

export function getDubheDepJS() {
  return resolve(root, 'dubhe.dep.js')
}

export function isExist(p: string) {
  return fse.existsSync(p)
}

export async function downloadFile(url: string, output: string) {
  try {
    const { data: code } = await axios.get(url)
    await fse.outputFile(
      output,
      typeof code === 'string' ? code : JSON.stringify(code),
    )
  }
  catch (e) {
    log(`get remote content fail--${url}`, 'red')
    process.exit(1)
  }
}

export async function installProjectCache(baseUrl: string, files: string[], project: string) {
  for (const file of files) {
    downloadFile(`${baseUrl}/core/${file}`, removeHash(getLocalPath(project, file)))
    log(`download file from ${baseUrl}/core/${file}`, 'grey')
  }
}

export async function installProjectTypes(baseUrl: string, project: string) {
  try {
    const typesFiles = await getRemoteContent(`${baseUrl}/types/types.json`)

    for (const file of typesFiles) {
      const typesPath = getTypePathInCache(project, file)
      downloadFile(`${baseUrl}/types/${file}`, typesPath)
    }
  }
  catch (e) {
    log('remote project doesn\' generate types.json', 'red')

    process.exit(0)
  }
}

