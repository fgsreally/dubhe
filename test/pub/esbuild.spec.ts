import { describe, expect, it } from 'vitest'
import { getDistFiles, getFileContent, isExist } from '../utils'
describe('[esbuild] publish module ', () => {
  it('core files output', async () => {
    const files = await getDistFiles('esbuild-pub/.dubhe/core')
    expect(files.length).toBe(5)
    const dubheList = JSON.parse(await getFileContent('esbuild-pub/.dubhe/core/dubheList.json'))
    delete dubheList.timestamp
    expect(dubheList).toMatchSnapshot()
  })

  it('types files output', async () => {
    expect(isExist('esbuild-pub/.dubhe/types/types.json')).toBeTruthy()
  })

  it('source files output', async () => {
    expect(isExist('esbuild-pub/.dubhe/source/src')).toBeTruthy()
  })
})
