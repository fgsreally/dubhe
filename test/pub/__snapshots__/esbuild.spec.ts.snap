// Vitest Snapshot v1

exports[`[esbuild] publish module  > core files output 1`] = `
{
  "alias": [
    {
      "name": "app",
      "url": "App.dubhe-esbuildpub",
    },
    {
      "name": "hello",
      "url": "HelloWorld.dubhe-esbuildpub",
    },
    {
      "name": "module",
      "url": "module.dubhe-esbuildpub",
    },
  ],
  "bundleGraph": {
    "App": [
      "chunk.dubhe-esbuildpub.js",
    ],
    "HelloWorld": [
      "chunk.dubhe-esbuildpub.js",
    ],
    "chunk": [],
    "module": [],
  },
  "entryFileMap": {
    "app": "./src/App.vue",
    "hello": "./src/HelloWorld.vue",
    "module": "./src/module.ts",
  },
  "externals": [
    "vue",
  ],
  "files": [
    "remoteEntry.js",
    "HelloWorld.dubhe-esbuildpub.js",
    "App.dubhe-esbuildpub.js",
    "chunk.dubhe-esbuildpub.js",
    "module.dubhe-esbuildpub.js",
  ],
  "from": "esbuild",
  "importsGraph": {
    "vue": [
      "defineComponent",
      "toDisplayString",
      "createElementVNode",
      "createVNode",
      "Fragment",
      "openBlock",
      "createElementBlock",
      "pushScopeId",
      "popScopeId",
      "*",
      "default",
    ],
  },
  "sourceGraph": {
    "./src/App.vue": [
      "src/test.ts",
      "src/App.vue",
    ],
    "./src/HelloWorld.vue": [],
    "./src/module.ts": [
      "src/module.ts",
    ],
  },
  "version": "0.0.0",
}
`;

exports[`[esbuild] publish module  > types files output 1`] = `6`;
